### PR TITLE
add file read function for datamovement library use

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

**Review by LarryOsterman — COMMENTED** [review]
Why not use CreateFile2 in both desktop and UWP?

**Comment by Jinming-Hu** [comment]
> Why not use CreateFile2 in both desktop and UWP?

Well this is not related to the change in this PR.

There's a history. We used `CreateFileW` for all platforms first, then we found it somewhat didn't work on UWP, so we start using `CreateFile2` for UWP. Another reason is that `CreateFile2` is only available since Windows 8.

**Comment by LarryOsterman** [comment]
FWIW, whenever you submit a pull request, you're going to get feedback on existing code as well as the new code. You never know what is going to catch the eye of the reviewer (especially since this is the first time I've taken a look at this code, and I always want to understand the code I review).

And to be fair, a lot of my questions can/should be handled with comments :). It also means that nosy code reviewers don't have to ask awkward questions about stuff that isn't a part of the pull request :).


Btw, you should seriously consider if `_get_osfhandle` can allow you to use the `open()` API for both Linux and Windows - that might enable you to simplify a bunch of your code (with the caveat that I don't remember if the default C locale on Windows is UTF-8 or not - if it is not, then to be 100% correct, I believe that you have no choice but to use the Win32 unicode APIs to open the file (if you're not familiar with how filesystem globalization works in Windows, I can explain this in more detail over email, but the bottom line is that Windows filesystem access is REALLY complicated if you care about globalization in the file names you are opening)).

**Comment by Jinming-Hu** [comment]
> Btw, you should seriously consider if _get_osfhandle can allow you to use the open() API for both Linux and Windows - that might enable you to simplify a bunch of your code (with the caveat that I don't remember if the default C locale on Windows is UTF-8 or not - if it is not, then to be 100% correct, I believe that you have no choice but to use the Win32 unicode APIs to open the file (if you're not familiar with how filesystem globalization works in Windows, I can explain this in more detail over email, but the bottom line is that Windows filesystem access is REALLY complicated if you care about globalization in the file names you are opening)).

The main functionality of these two classes is to read/write file from different offset in a thread-safe manner (without depending on the internal offset associated with the fd or file handle). The way to do this on Windows is OVERLAPPED I/O with `ReadFile()` and `WriteFile()`. If I cannot get rid of these two Windows APIs, I feel it's better to also use Windows APIs to open and close the files, to have a consistent styles. Another problem with `open()` or `fopen()` is that we cannot specify share mode.

**Comment by LarryOsterman** [comment]
> > Btw, you should seriously consider if _get_osfhandle can allow you to use the open() API for both Linux and Windows - that might enable you to simplify a bunch of your code (with the caveat that I don't remember if the default C locale on Windows is UTF-8 or not - if it is not, then to be 100% correct, I believe that you have no choice but to use the Win32 unicode APIs to open the file (if you're not familiar with how filesystem globalization works in Windows, I can explain this in more detail over email, but the bottom line is that Windows filesystem access is REALLY complicated if you care about globalization in the file names you are opening)).
> 
> The main functionality of these two classes is to read/write file from different offset in a thread-safe manner (without depending on the internal offset associated with the fd or file handle). The way to do this on Windows is OVERLAPPED I/O with `ReadFile()` and `WriteFile()`. If I cannot get rid of these two Windows APIs, I feel it's better to also use Windows APIs to open and close the files, to have a consistent style. Another problem with `open()` or `fopen()` is that we cannot specify share mode.

IMHO you should minimize the number of platform specific dependencies. That is *especially* true w.r.t. anything that interacts with anything that has to do with filenames (the instant you allow filenames that aren't in the US-ASCII range, things get incredibly complicated). If you stick with the C runtime library, the problem ceases to be your problem and becomes the C runtime libraries problem.

But my core issue with this set of changes is the almost complete lack of documentation. There are some important considerations being made here that drive the choices you made and none of them are explained. It is important for future maintainers (and 3rd party contributors) that they be well documented.

The very fact that I had to ask all these questions indicates that the existing documentation in the code isn't sufficient to explain what is happening here.

**Comment by Jinming-Hu** [comment]
I created an issue https://github.com/Azure/azure-sdk-for-cpp/issues/3619 to track docs problem

**Comment by katmsft** [comment]
Do we have unit test coverage for the added function?

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (EmmaZhu) [inline sdk/storage/azure-storage-common/src/file_io.cpp:134]: Any way to add more detailed error messages, or at lease include the error code into the message?

**Comment by Jinming-Hu** [comment]
> Do we have unit test coverage for the added function?

No, we don't test internal methods. Internal stuff is tested via public methods that use this thing internally. For this case, the Read() function is used in datamovement library, so it will be covered there.

**Comment by katmsft** [comment]
Seems to me an external method that is called in another library, a test coverage would avoid the functionality issue or break in the future.
Anyway, it's your call. Approving to unblock.

**Review by katmsft — APPROVED** [review]

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp:134]: Yeah, we could do better. To be honest, I don't expect the operation will fail very often. I mean, you've opened the file successfully. Under which circumstances will the read operation fail? Maybe when you shrink the file while the library is trying to read it? But thinking from a different angle, it is kind of obvious if you edit a file that is being uploaded, then the upload is failed. A more detailed error message won't help a lot in this case maybe?

**Review by LarryOsterman — CHANGES_REQUESTED** [review]

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp:58]: `open()` and `fopen()` are both deprecated on Windows in favor of the security enhancement version of them (the ones with `_s`). So I cannot get rid of the conditional compiling anyway. Moreover, it's more difficult to specify share mode with C-runtime API. So I guess I'll add some comments here so that it's easier to understand.

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp:35]: I'm looking at the code snippet at the end of the blob. It seems to me you also filled the buffer with some initial data.

```C++
   std::wstring UnicodeStringFromAnsiString(_In_ const std::string &ansiString)

   {

       const std::string::size_type limit = std::numeric_limits<int>::max();

       assert(ansiString.size() < limit);

       const int ansiByteSize = static_cast<int>(ansiString.size());

       const auto wideCharSize =

           ::MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, ansiString.c_str(), ansiByteSize,

                                 nullptr, 0);

       std::wstring returnValue(wideCharSize, L'');                     // <----------------- Here

       const auto finalCharSize =

           ::MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, ansiString.c_str(), ansiByteSize,

                                 &returnValue[0], wideCharSize);

       assert(finalCharSize == wideCharSize);

       return returnValue;

   }
```

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp:35]: Anyway, I believe this is a very trivial optimization, and I don't think it's necessary to do it at all in a network application.

**Pull request closed without merging**
